COS-4: Creating Custom fields for Contact and Contribution Entities with installation of COS extension

awg-github opened the pull request:
# Scope
1. CiviCRM Odoo Synchronisation extension installation.
2. The installation of extension create the custom fields (using XML method):
    odoo_partner_sync_information for Contact Entity
    odoo_invoice_sync_information for Contribution Entity
    purchase_order for Contribution Entity
3. When the extension is installed: 	
   The "Sync Status" field for each contact record is filled with "Awaiting Sync". The "Action to Sync" field for each contact record is filled with "Create".
   The "Sync Status" field for each contribution record is filled with "Awaiting Sync". The "Action to Sync" field for each contribution record is filled with "Create".

# After
![image](https://user-images.githubusercontent.com/36959503/37216323-2c2c1a6c-23c3-11e8-87d1-0815fdd999c9.png)
![image](https://user-images.githubusercontent.com/36959503/37216326-2f1ca8c2-23c3-11e8-88d0-6997049e0ace.png)
![image](https://user-images.githubusercontent.com/36959503/37216330-3167245e-23c3-11e8-9668-a854f91c4e3d.png)

awg-github committed:
COS-4: Creating Custom fields for Contact and Contribution Entities with installation of COS extension

omarabuhussein commented:
hi @awg-github , I will review this on Monday if that's okay

omarabuhussein commented on CRM/Odoocivicrmsync/Upgrader.php:
```diff
@@ -0,0 +1,94 @@
+   *
+   * @throws \CiviCRM_API3_Exception
+   */
+  public function postInstall() {
```
please move postInstall method above uninstall method.

omarabuhussein commented on CRM/Odoocivicrmsync/Upgrader.php:
```diff
@@ -0,0 +1,94 @@
+   * @throws \CiviCRM_API3_Exception
+   */
+  public function postInstall() {
+    $contactIds = $this->getAllContactId();
```
While it is better to use API instead of SQL when dealing with CiviCRM core entities (like custom fields), but in this case it is inefficient to do that, Imagine the site contain thousands of contacts, I would instead replace this whole implementation with a single SQL queries : 

```php
INSERT INTO odoo_partner_sync_information(entity_id, replace_this_with_the_field_1_name, replace_this_with_the_field_2_name) 
SELECT id, 1, 1 FROM 	civicrm_contact;
```

omarabuhussein commented on CRM/Odoocivicrmsync/Upgrader.php:
```diff
@@ -0,0 +1,94 @@
+   * @throws \CiviCRM_API3_Exception
+   */
+  public function uninstall() {
+    $this->deleteGroup('OptionGroup', 'odoo_sync_status');
```
instead of doing that, you can the following : 


```php
public function uninstall() {
  $this->deleteExtensionOptionGroups();
  $this->deleteExtensionCustomGroups();
}

private function deleteExtensionOptionGroups(){
  $optionGroupsToDelete = ['odoo_sync_status', 'odoo_partner_action_to_sync', 
  'odoo_invoice_action_to_sync'];
  foreach($optionGroupsToDelete as $optionGroupName) {
    $this->deleteOptionGroup($optionGroupName);
  }
}

private function deleteExtensionCustomGroups(){
  $customGroupsToDelete = ['odoo_invoice_sync_information', 'odoo_partner_sync_information', 
  'purchase_order'];
  foreach($customGroupsToDelete as $customGroupName) {
    $this->deleteCustomGroup($customGroupName);
  }
}
...etc
```

omarabuhussein commented on CRM/Odoocivicrmsync/Upgrader.php:
```diff
@@ -0,0 +1,94 @@
+   */
+  private function deleteGroup($groupType, $fieldIdentifier, $searchBy = 'name') {
+    civicrm_api3($groupType, 'delete', [
+      'id' => $this->getIdByName($groupType, $fieldIdentifier, $searchBy),
```
you can use chained API call , see my comment here : https://github.com/compucorp/uk.co.compucorp.manualdirectdebit/pull/7#discussion_r173499361

omarabuhussein commented on xml/customFields_install.xml:
```diff
@@ -0,0 +1 @@
            <profile_group_name>purchase_order</profile_group_name>
        </ProfileJoin>
    </ProfileJoins>
</CustomData>
```
this line endings in this  file look messed up, please use Unix line ending to fix it.

also there are <ProfileGroups> tag used, I never used this tag before but my understanding is that it is related to civicrm profiles, but I cannot see that it is part of ticket specs, can you please elaborate more and if it reallyreally needed to be added here ?

omarabuhussein reviewed (changes requested):
@awg-github 

I left some  inline notes, but here are another 2 : 

1- The extension repo name is : uk.co.compucorp.odoosync, so the name used in the code should be odoosync, but you used odoocivicrmsync instead, so please rename the extension name in code to be similar to the repo name to use odoosync.

2- To make it easier for who review the PR or for anyone who is reviewing git history anytime later to make meaningful commits separate, it is maybe too late for this PR but for example you could have created the extension skeleton files in one commit, added the custom fields xml file in another commit, implemented the uninstall in another commit and implemented postInstall in another commit instead of doing all of them in one commit. So please pay attention for that in next PRs.

awg-github committed:
COS-4: Rename extension name to ‘odoosync’. Refactor method ‘updateContactCustomData’ and  install/uninstall extension. Fix custom field xml file.

awg-github commented:
@omarabuhussein please check my fixes.

- Renamed extension name to ‘odoosync’. 
- Refactored method ‘updateContactCustomData’ and  install/uninstall extension. 
- Fixed custom field xml file.

omarabuhussein commented on CRM/Odoosync/Upgrader.php:
```diff
@@ -0,0 +1,94 @@
+  /**
+   *  Update contact custom data
+   */
+  private function updateContactCustomData() {
```
updateContactCustomData name is too generic, the name should reflect what the method is about, in this case a name like : 

initializeContactsSyncInformation()

makes more sense.

omarabuhussein commented on CRM/Odoosync/Upgrader.php:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  /**
+   *  Update contact custom data
```
also please update the docblock to reflect what the method is about, this description is too generic

omarabuhussein commented on CRM/Odoosync/Upgrader.php:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  /**
+   * Delete all option group from ext
```
according to drupal standrads : https://www.drupal.org/docs/develop/coding-standards/api-documentation-and-comment-standards

 If a file docblock summary begins with a verb, it should be in third person singular present tense.

so in this case it should be "Deletes" instead of "Delete", please apply this for all methods.


also no need to use abbreviations here, so use extension instead of ext.

in general a more clear description would be : 

**Deletes all the option groups created by the extension**

omarabuhussein commented on CRM/Odoosync/Upgrader.php:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  /**
+   * Delete all custom group from ext
```
as before, replace  this with : Deletes all the custom groups created by the extension

omarabuhussein commented on CRM/Odoosync/Upgrader.php:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  /**
+   * Delete custom group from ext
```
replace the description with : 

Deletes the specified custom group

omarabuhussein commented on CRM/Odoosync/Upgrader.php:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  /**
+   * Delete option group from ext
```
replace the description with : 

Deletes the specified option group

omarabuhussein commented on CRM/Odoosync/Upgrader.php:
```diff
@@ -0,0 +1,94 @@
+  /**
+   * Delete option group from ext
+   *
+   * @param $optionGroupName
```
please add the param type here and in all other methods, so : 

@param string $optionGroupName

omarabuhussein reviewed (changes requested):
left some notes, also two things : 

1- the xml file is still looks broken (check github diff) .

2-  Also please as I said before, when doing more than one thing then separate it into multiple commits, otherwise it is pain to review it or seeing what changed . 

For example you could have just renamed the extension in one commit, refactored the upgrader in another commit and fixed the xml file in another one.

awg-github committed:
COS-4: Method's comment refactoring

awg-github committed:
COS-4: Set Unix line separator.

awg-github commented:
@omarabuhussein done. 
Please, check my fixes

omarabuhussein commented on CRM/Odoosync/Upgrader.php:
```diff
@@ -79,9 +79,10 @@ private function deleteCustomGroup($customGroupName) {
 
   /**
-   *  Update contact custom data
+   *  Initialize contacts sync information
```
in the description use : **Initializes** instead of **Initialize**

omarabuhussein commented on CRM/Odoosync/Upgrader.php:
```diff
@@ -79,9 +79,10 @@ private function deleteCustomGroup($customGroupName) {
+   *  Set custom field "action_to_sync" and "sync_status" default values
    */
-  private function updateContactCustomData() {
+  private function initializeContactsSyncInformation() {
```
sorry I forget to mention this before, instead of using values directly (e.g 1,1) in the query below directly you should fetch the values from the API instead then use them in the query.

awg-github committed:
COS-4: Fix 'initializeContactsSyncInformation' metod.

awg-github commented:
@omarabuhussein done

omarabuhussein commented on CRM/Odoosync/Upgrader.php:
```diff
@@ -79,17 +79,44 @@ private function deleteCustomGroup($customGroupName) {
+      'is_default' => 1,
+    ]);
+
+    return (int) $value;
```
shouldn't you do : 

return (int) $value['values'][0]['value'];

instead of 

return (int) $value;

awg-github replied:
@omarabuhussein fixed

awg-github committed:
COS-4: fixed return value

awg-github reviewed (commented):
@omarabuhussein fixed

omarabuhussein commented:
approved @awg-github
